---
title: "ExpHunterSuite: Differential Expression Report"
output:
  html_document:
    toc: true
    toc_float: true
    df_print: paged
    fig_width: 12
---

<style type='text/css'>
    body .main-container {
        max-width: 90%;

    }

    .plot_real_size {
    	overflow: scroll;
  		max-height: 600px;
  	}

    .plot_real_size img{
    	max-width: none;
    	max-height: none;
    }

    embed {
   
  }   
</style>

## **Details of the input data**

### **First group of samples (to be referred to as control in the rest of the report)**
```{r control_table, echo = FALSE, warning=FALSE}
	knitr::kable(index_control_cols, col.names="Sample Names:")
```

### **Second group of samples (to be referred to as treatment in the rest of the report)**
```{r treatment_table, echo = FALSE, warning=FALSE}
	knitr::kable(index_treatmn_cols, col.names="Sample Names:")

```

```{r plots_configuration, echo = FALSE, warning=FALSE}
	# this counter is defined for plotting graphs with plot_in_div function. If not defined as a number, plots can be mixed.
	plot_counter <- 1 
```
Note: A positive log fold change shows higher expression in the treatment group; a negative log fold change represents higher expression in the control group.

## **Data quality control (QC)**

### **Correlation between samples:**

Here we show scatterplots comparing expression levels for all genes between the different samples, for i) all controls, ii) all treatment samples and iii) for all samples together.

These plots will only be produced when the total number of samples to compare within a group is less than or equal to 10.

```{r corr_plot_control, echo = FALSE, fig.height=12, fig.width=12, eval=length(index_control_cols) <= 10, results='asis', warning = FALSE}

cat("### **Correlation between control samples:**
Replicates within the same group tend to have Pearson correlation coefficients >= 0.96. Lower values may indicate problems with the samples.
")

PerformanceAnalytics::chart.Correlation(raw_filter[index_control_cols], histogram=TRUE, pch=19, log="xy")

```

```{r corr_plot_treatment, echo = FALSE, fig.height=12, fig.width=12, eval=length(index_treatmn_cols) <= 10, results='asis', warning = FALSE}

cat("### **Correlation between treatment samples:**
Replicates within the same group tend to have Pearson correlation coefficients >= 0.96. Lower values may indicate problems with the samples.
")

PerformanceAnalytics::chart.Correlation(raw_filter[index_treatmn_cols], histogram=TRUE, pch=19, log="xy")


```


```{r corr_plot_all , echo = FALSE, fig.height=12, fig.width=12, eval=length(index_control_cols) + length(index_treatmn_cols) <= 10, results='asis', warning = FALSE}


cat("### **Correlation between samples: All vs all replicates**

Correlation coefficients tend to be slightly higher between replicates from the same group than between replicates from different groups. If this is not the case, it may indicate mislabelling or other potential issues.

")

PerformanceAnalytics::chart.Correlation(raw_filter, histogram=TRUE, pch=19, log="xy")

```



### **Heatmap and clustering showing correlation between replicates**
**BROWN: higher correlation; YELLOW: lower**

```{r heatmap_corr, eval=!"externalDEA" %in% names(all_counts_for_plotting), echo = FALSE, fig.height=12, fig.width=12}
col<- colorRampPalette(c("#FFFFCC", "#FD8D3C", "#800026"))(20)
res <- cor(raw_filter)
gplots::heatmap.2(x = res, col = col, symm = TRUE, margins = rep(max(nchar(colnames(res))), 2), scale = "none", trace = "none")
```

### **Principal Component Analysis**
This is a PCA plot of the count values following rlog normalization from the DESeq2 package:

```{r pca_plot, eval="DESeq2" %in% names(all_counts_for_plotting), echo = FALSE, results = "asis", warning = FALSE, message = FALSE}
coldata_df <- data.frame(cond = design_vector,
                           each = colnames(raw_filter),
                           row.names = colnames(raw_filter), stringsAsFactors=TRUE)

dds <- DESeq2::DESeqDataSetFromMatrix(countData = raw_filter, colData = coldata_df, design = ~ cond)

if(nrow(raw_filter) < 2000 || sum((rowSums(raw_filter) / ncol(raw_filter)) > 5) < 1000 ) {
  rld <- DESeq2::rlog(dds, blind=TRUE)
} else {
	rld <- DESeq2::vst(dds)
}

pcaData <- DESeq2::plotPCA(rld, intgroup=c("cond"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
pca_plot <- ggplot2::ggplot(pcaData, ggplot2::aes(x = PC1, y = PC2, color = name, shape = cond)) +
  ggplot2::geom_point(size =3) +
  ggrepel::geom_text_repel(ggplot2::aes(label=name),vjust="inward",hjust="inward") +
  ggplot2::xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ggplot2::ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  ggplot2::coord_fixed() +
  ggplot2::theme(legend.position="none")
plot_counter <- plot_in_div(pca_plot)
# +theme(legend.position="bottom", legend.box = "horizontal")

# ggplot2::ggplot(pcaData, ggplot2::aes(x = PC1, y = PC2, color = name, shape = cond)) +
#   ggplot2::geom_point(size =3) +
#   ggplot2::theme(legend.position = "bottom") +
#   ggplot2::xlab(paste0("PC1: ", percentVar[1], "% variance")) +
#   ggplot2::ylab(paste0("PC2: ", percentVar[2], "% variance")) +
#   ggplot2::coord_fixed() 


```
```{r pca_plot_no, eval=! "DESeq2" %in% names(all_counts_for_plotting), echo = FALSE}
cat("\n\n## DESeq2 not run so plot not shown\n\n")
```

The samples are shown in the 2D plane and distributed by their first two principal components. This type of plot is useful for visualizing the overall effect of experimental covariates and batch effects. It is also useful for identifying outlier samples. Control and treatment samples respectively may cluster together.

### **Visualizing normalization results**

These boxplots show the distributions of count data before and after normalization (shown for normalization method `r names(all_data_normalized)[1]`):

Representation of cpm unfiltered data:
```{r density_distribution_cpm, echo = FALSE, warning=FALSE}
cpm_no_zeros <- cpm_table[! rowSums(cpm_table) == 0, ]
log2cpm <- log2(cpm_no_zeros + 1)
cpm_long <- reshape2::melt(log2cpm, measure.vars=colnames(log2cpm))

ggplot2::ggplot(cpm_long, ggplot2::aes(x = value, colour=Var2)) + ggplot2::geom_density() +
ggplot2::xlab(expression(log[2](cpm + 1))) +
ggplot2::geom_vline(xintercept = log2( opt$reads + 1))

```


Before normalization:

```{r density_distribution_raw, echo = FALSE, warning=FALSE}
log2raw_counts <- log2(raw_filter + 1)
raw_long <- reshape2::melt(log2raw_counts, measure.vars=colnames(log2raw_counts))

ggplot2::ggplot(raw_long, ggplot2::aes(x = value, colour=variable)) + ggplot2::geom_density() +
ggplot2::xlab(expression(log[2](count + 1)))

```


After normalization:

```{r density_distribution_norm, echo = FALSE, warning=FALSE}
log2norm_counts <- log2(all_data_normalized[[1]] + 1)
norm_long <- reshape2::melt(log2norm_counts, measure.vars=colnames(log2norm_counts))

ggplot2::ggplot(norm_long, ggplot2::aes(x = value, colour=variable)) + ggplot2::geom_density() +
ggplot2::xlab(expression(log[2](count + 1)))

```

#############################################################################
#############################################################################
########################## NOOB ALERT @alvaro ###############################
#############################################################################
#############################################################################

## **Count metrics by sample ranks**

```{r coverage_plot_head, echo = FALSE, warning = FALSE, results='asis'}
	if (!is.null(library_sizes)) {
		cat('### **Sample rank versus total reads**')
	}else{
		cat('### **Sample rank versus total counts**')
	}
```

```{r coverage_plot, echo = FALSE, warning = FALSE}
    if (!is.null(library_sizes)) {
        xtitle="Sample Rank (total reads)"
    } else {
        xtitle="Sample Rank (total counts)"
    }
    ggplot2::ggplot(coverage_df, ggplot2::aes(x = sample_rank, y = total_counts)) +
        ggplot2::geom_point(size = 3) +
        cowplot::theme_cowplot() +
        cowplot::background_grid() +
        ggplot2::labs(title = "Read Counts", x=xtitle, y = "Reads Counted") +
        ggplot2::ylim(c(0,NA)) +
        ggplot2::scale_color_brewer(palette="Dark2")
```
```{r coverage_plot_text, echo = FALSE, warning = FALSE, results='asis'}
	if (!is.null(library_sizes)) {
		cat('\n\n Sample rank is the position a sample holds after sorting by total reads')
	}else{
		cat('\n\n Sample rank is the position a sample holds after sorting by total counts')
	}
```

```{r coverage_plot_frac_head, eval = !is.null(coverage_df$counted_frac),  echo = FALSE, warning = FALSE, results='asis'}
	cat('### **Fraction counts read sorted by rank**')
```

```{r coverage_plot_frac, eval = !is.null(coverage_df$counted_frac),  echo = FALSE, warning = FALSE}
    ggplot2::ggplot(coverage_df, ggplot2::aes(x = sample_rank, y = counted_frac)) +
        ggplot2::geom_point(size = 3) +
        cowplot::theme_cowplot() +
        cowplot::background_grid() +
        ggplot2::labs(title = "Read Counts", x="Sample Rank (total reads)", y = "% counts read") +
        ggplot2::ylim(c(0,NA)) +
        ggplot2::scale_color_brewer(palette="Dark2")
```

```{r coverage_plot_frac_text, eval = !is.null(coverage_df$counted_frac),  echo = FALSE, warning = FALSE, results='asis'}
	cat('Counted fraction is the percentage of reads that map to a gene (only available if total reads provided)')
```

## **Statistics of expressed genes**

```{r exp_genes_plot, echo = FALSE, warning = FALSE}
		colors <- c("Unique" = "blue", "Union" = "red", "Intersection" = "green")
		ggplot2::ggplot(data = exp_genes_df, ggplot2::aes(x = count_rank)) +
		    ggplot2::geom_line(ggplot2::aes(y = expressed_genes, color = "Unique")) +
		    ggplot2::geom_line(ggplot2::aes(y = union_expressed_genes, color = "Union")) +
		    ggplot2::geom_line(ggplot2::aes(y = inters_expressed_genes, color = "Intersection")) +
		    ggplot2::labs(x = "Sample rank",
		                  y = "Number of genes",
		                  color = "Legend") +
		    ggplot2::expand_limits(y=0) +
		    ggplot2::scale_color_manual(values = colors)
```

Samples are ranked by total expressed genes. Union of expressed genes represents the cumulative total expressed genes (sum of
all genes expressed in any sample up to current sample, expected to increase with sample rank). Intersection of expressed genes
represents the cumulative intersection of expressed genes (sum of genes expressed in all samples up to current sample, expected
to decrease with sample rank).

### **Mean count distribution by filter**

```{r mean_counts_plot, echo = FALSE, warning = FALSE}
    ggplot2::ggplot(mean_counts_df, ggplot2::aes(x = counts, col = filter)) +
	ggplot2::geom_density(ggplot2::aes(y=..count..)) +
	ggplot2::scale_x_log10() +
	ggplot2::labs(x = "Mean counts per gene", y = "Frequency") +
    ggplot2::guides(col = ggplot2::guide_legend(title = NULL)) +
    ggplot2::scale_color_brewer(palette = "Paired") +
    cowplot::theme_cowplot() +
    ggplot2::theme(legend.position = "top",
          legend.justification="center",
          legend.background = ggplot2::element_rect(color = NA))
```

This plot represents the mean counts distribution per gene, classified by filters

## Gene counts variance distribution

Variance of gene counts across samples are represented. Genes with lower variance than selected threshold (dashed grey line) were filtered out. 

```{r variance_dist, echo = FALSE, eval=!grepl("F", opt$modules), warning = FALSE, message = FALSE}

gene_count_var <- data.frame(variance =var_filter[["variance_dis"]])

ggplot2::ggplot(gene_count_var, ggplot2::aes(x = variance))+
ggplot2::geom_density() + 
ggplot2::geom_vline(xintercept = var_filter[["thr"]], linetype="dashed", 
                color = "grey")
```

```{r DESeq2_all_norm_head, eval="DESeq2" %in% names(all_counts_for_plotting), echo=FALSE, results='asis'}

cat("### **Samples differences by all counts normalized**:
		
All counts were normalizated by DESeq2 algorithm. This count were scaled by log10 and plotted in a heatmap.")
```

```{r heatmap_all_norm_plot, eval="DESeq2" %in% names(all_counts_for_plotting), echo = FALSE, fig.height=12, fig.width=12}
	
		deseq_normalized_data <- all_data_normalized[['DESeq2']]
	if (nrow(all_data_normalized[['DESeq2']]) <= 50000) { 
		deseq_normalized_data_mtx <- log10(data.matrix(deseq_normalized_data) + 1)

		plot <- gplots::heatmap.2(deseq_normalized_data_mtx, col = col, labRow = FALSE, dendrogram = "column", trace = "none", srtCol=45,  adjCol = c(1,1))
	} else {
		cat("\n\nThis plot can not be generated because more than 50 000 genes have passed the filters and it can cause memory problems.\n\n")
	}

```

```{r top_n_expression, results='asis', echo=FALSE, warning=FALSE, message=FALSE}
	cat("\n\n### **Percentages of reads per sample mapping to the most highly expressed genes **\n\n")
	make_top_n_expression_table(count_data=raw_filter, n=5)
```

## **DEgenes Hunter results**

### **Gene classification by DEgenes Hunter**
DEgenes Hunter uses multiple DE detection packages to analyse all genes in the input count table and labels them accordingly:

* **Filtered out:** Genes discarded during the filtering process as showing no or very low expression.
* **Prevalent DEG:** Genes considered as differentially expressed (DE) by at least `r opt$minpack_common` packages, as specified by the `minpack_common` argument.
* **Possible DEG:** Genes considered DE by at least one of the DE detection packages.
* **Not DEG:** Genes not considered DE in any package.

This barplot shows the total number of genes passing each stage of analysis - from the total number of genes in the input table of counts, to the genes surviving the expression filter, to the genes detected as DE by one package, to the genes detected by at least `r opt$minpack_common` packages.

```{r qc,echo = FALSE}
# Create genes barplot
barplot_df <- data.frame(total=c(
		sum(DE_all_genes[, "genes_tag"] %in% c("PREVALENT_DEG", "POSSIBLE_DEG", "NOT_DEG", "FILTERED_OUT")),
		sum(DE_all_genes[, "genes_tag"] %in% c("PREVALENT_DEG", "POSSIBLE_DEG", "NOT_DEG")),
		sum(DE_all_genes[, "genes_tag"] %in% c("PREVALENT_DEG", "POSSIBLE_DEG")),
		sum(DE_all_genes[, "genes_tag"] %in% c("PREVALENT_DEG"))
		),
	  gene_type=c("Total genes", "Genes passing expression filter", "DEGs considered possible or prevalent", "Prevalent DEGs")
	)
	barplot_df[,"gene_type"] <- factor(barplot_df[,"gene_type"], levels = barplot_df[,"gene_type"][order(barplot_df[,"total"])]) # Reorder to plot from most to least

# Plot genes barplot
pgenes <- ggplot2::ggplot(barplot_df, ggplot2::aes(gene_type, total)) +
		  ggplot2::ylab("Number of genes") + 
		  ggplot2::xlab("") +
		  ggplot2::geom_bar(position="dodge", stat="identity", fill=c("#000034", "red", "orange", "blue"), show.legend=FALSE) + ggplot2::coord_flip() + 
		  ggplot2::geom_text(ggplot2::aes(label = total, y= total + 1500))+ 
		  ggplot2::theme(text = ggplot2::element_text(face="bold", size=10))

plot(pgenes)
```

### **Package DEG detection stats**
This is the Venn Diagram of all possible DE genes (DEGs) according to at least on of the DE detection packages employed:

```{r Venn_d, echo = FALSE, fig.height=12, fig.width=12}
sets <- list()
for(DEG_pack in DEG_pack_columns){
	pack <- unlist(strsplit(DEG_pack, split='_'))[1] # Get the name of the package
	sets[[pack]] <- row.names(DE_all_genes)[which(DE_all_genes[,DEG_pack] == TRUE)] # 
}
items <- unique(unlist(sets))

# Per each item, check sets
belonging <- unlist(lapply(sets, function(set){items %in% set}))
belonging <- matrix(as.numeric(belonging),ncol = length(sets))
# Add set names
colnames(belonging) <- names(sets)
# Calculate intersections
intersections <- limma::vennCounts(belonging)
intersections[1,"Counts"] <- NA
# Plot
limma::vennDiagram(intersections, circle.col = grDevices::rainbow(length(sets)))

```

```{r setup_logFCplots, echo = FALSE}
var_thr <- 0.01
```

```{r logFCplots, echo = FALSE, eval=(!grepl("F", opt$modules) && sum(grepl("logFC_", colnames(DE_all_genes))) > 1), results = "asis"}

  cat(paste0("\n### **Plot showing variability between different DEG detection methods in terms of logFC calculation**"))
  cat(paste0("\nThis graph shows logFC calculated (y-axis) for each package (points) and gene (x-axis). Only genes with variability over ", var_thr, " will be plotted. This representation allows to user to observe the behaviour of each DE package and see if one of them has atypical results.\n"))
  cat(paste0("\nIf there are no genes showing sufficient variance in estimated logFC accross methods, no plot will be produced and a warning message will be given.\n\n"))

  pp <- ht2logFCPlot(DE_all_genes, var_filter = var_thr, title = paste0("(variance threshold >= ",var_thr,")"))
	print(pp)
```


### **FDR gene-wise benchmarking**

Benchmark of false positive calling:

Boxplot of FDR values among all genes with an FDR <= 0.05 in at least one DE detection package

```{r FDR_plot, echo = FALSE}
df <- DE_all_genes[DE_all_genes[, "genes_tag"] == "PREVALENT_DEG",]
df <- df[final_FDR_names]

df$gene_names <- row.names(df)
df_melted_FDR <- reshape2::melt(df, id.vars = "gene_names", measure.vars = final_FDR_names)
df_melted_FDR[,"package_name"] <- unlist(lapply(strsplit(as.character(df_melted_FDR[, "variable"]), split='_'), function(x) x[2])) # Ugly way to remove FDR_ from names

if(dim(df)[1] == 0){
	cat("No Prevalent DEGs found, Bar charts of FDR values for prevalent genes cannot be shown")
} else { 

p_seguros_Int <- ggplot2::ggplot(df_melted_FDR, ggplot2::aes(x = package_name, y = value, color = package_name)) +
    			 ggplot2::geom_boxplot(outlier.colour = grDevices::rgb(0, 0, 0, 0)) + 
		         ggplot2::theme_bw(base_size = 30) + 
		         ggplot2::geom_point(position = ggplot2::position_jitter(w = 0.1), color = "grey50", size = 1) + 
		         ggplot2::geom_hline(ggplot2::aes(yintercept = opt$p_val_cutoff)) + 
		         ggplot2::ylab("1 - precision (FDR)") + 
		         ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 45, hjust = 1)) + 
		         ggplot2::xlab("") + 
		         ggplot2::scale_colour_discrete(guide = "none") + 
		         ggplot2::coord_cartesian(ylim = c(0, opt$p_val_cutoff))
plot(p_seguros_Int)
}
```

### **FDR Volcano Plot showing log 2 fold change vs. FDR**
The red horizontal line represents the chosen FDR threshold of `r opt$p_val_cutoff`. The black lines represent
other values.


```{r FDR_volcano, echo = FALSE, warning=FALSE}

volcano_DE_all_genes <- DE_all_genes
volcano_DE_all_genes <- volcano_DE_all_genes[volcano_DE_all_genes$genes_tag != "FILTERED_OUT",]

if (any(volcano_DE_all_genes$combined_FDR[!is.na(volcano_DE_all_genes$combined_FDR)] == 0)){
	min_FDR <- min(volcano_DE_all_genes$combined_FDR[volcano_DE_all_genes$combined_FDR != 0 &
																									 !is.na(volcano_DE_all_genes$combined_FDR)])
	volcano_DE_all_genes[volcano_DE_all_genes$combined_FDR == 0 &
											 !is.na(volcano_DE_all_genes$combined_FDR), "combined_FDR"] <- min_FDR
}

custom_shapes <- ifelse(volcano_DE_all_genes$genes_tag == "PREVALENT_DEG", 16,
													 ifelse(volcano_DE_all_genes$genes_tag == "POSSIBLE_DEG", 0, 3)) 
names(custom_shapes) <- volcano_DE_all_genes$genes_tag

evp <- EnhancedVolcano::EnhancedVolcano(volcano_DE_all_genes, 
	lab=rownames(volcano_DE_all_genes),  
	x="mean_logFCs", 
	y="combined_FDR",     
	boxedLabels = TRUE,     
	drawConnectors = TRUE,     
	pCutoff = opt$p_val_cutoff,     
	FCcutoff = opt$lfc,  
	cutoffLineType = 3,
  ylab = bquote(~-Log[10] ~ italic(FDR)),
  title = NULL,
  subtitle = NULL,
  legendLabels = c("NS", expression(Log[2] ~ FC), "FDR", expression(FDR ~ and
         ~ log[2] ~ FC)),
  caption=NULL,
  legendPosition = 'right',
  shapeCustom = custom_shapes
) 

fc_thresholds <- c(0.01, 0.05,0.1, 0.2)
all_thresholds <- unique(c(opt$p_val_cutoff, fc_thresholds))
mean_wt = data.frame(FDR = all_thresholds, nlFDR = -log10(all_thresholds))

evp <- evp +
ggplot2::geom_hline(data = mean_wt, ggplot2::aes(yintercept = nlFDR, linetype=factor(FDR)) , color = c("red", rep("black", times=nrow(mean_wt)-1)))
plot(evp)

```



The complete results of the DEgenes Hunter differential expression analysis can be found in the "hunter\_results\_table.txt" file in the Common_results folder

## **DE detection package specific results**

Various plots specific to each package are shown below:

```{r DESeq2_conditional_block_sample_rank_head, eval="DESeq2" %in% names(all_counts_for_plotting), echo=FALSE, results='asis'}

cat("### **DESeq2 size factor vs. sample rank**")
```

```{r DESeq2_conditional_block_sample_rank_plot, eval="DESeq2" %in% names(all_counts_for_plotting), echo=FALSE, warning = FALSE, message = FALSE}
	effective_lib_size <- DESeq2::sizeFactors(package_objects[['DESeq2']][['DESeq2_dataset']])
	effective_lib_size <- effective_lib_size[match(coverage_df$sample_ID, names(effective_lib_size))] 
	sample_rank <- coverage_df$sample_rank
	size_vs_rank <- data.frame(sample_rank = sample_rank,
								eff_lib_size = effective_lib_size)
	if (!is.null(library_sizes)) {
		xtitle="Sample Rank (total reads)"
	} else {
		xtitle="Sample Rank (total counts)"
    }

    ggplot2::ggplot(size_vs_rank, ggplot2::aes(x = sample_rank, y = eff_lib_size)) +
        ggplot2::geom_point(size = 3) +
        cowplot::theme_cowplot() +
        cowplot::background_grid() +
        ggplot2::labs(x=xtitle, y = "Effective library size \n(DESeq2 normalization factor)") +
        ggplot2::ylim(c(0,NA)) +
        ggplot2::scale_color_brewer(palette="Dark2")

```

```{r DESeq2_conditional_block_sample_rank_text, eval="DESeq2" %in% names(all_counts_for_plotting), echo=FALSE, results='asis'}

cat("The effective library size is the factor used by DESeq2 normalization algorithm for each sample. The effective library size must be dependent of raw library size.")
```

```{r DESeq2_conditional_block_ELS_head, eval="DESeq2" %in% names(all_counts_for_plotting), echo=FALSE, results='asis'}

cat("### **DESeq2 normalization effects**:

This plot compares the effective library size with raw library size")
```

```{r DESeq2_conditional_block_ELS_plot, eval="DESeq2" %in% names(all_counts_for_plotting), echo=FALSE, warning = FALSE, message = FALSE}
	effective_lib_size <- DESeq2::sizeFactors(package_objects[['DESeq2']][['DESeq2_dataset']])
	sums <- colSums(raw_filter)
	norm_bias <- data.frame(sums = sums, 
							eff_lib_size = effective_lib_size,
							Treat = design_vector)
	rownames(norm_bias) <- names(sums)

	plot <- ggplot2::ggplot(norm_bias, ggplot2::aes(x = sums, y = eff_lib_size, group = Treat)) + 
		ggplot2::geom_point(ggplot2::aes(shape = Treat, color = Treat, size = 8)) +
		ggplot2::geom_smooth(method = "lm", fill = NA, ggplot2::aes(color = Treat))+
		ggplot2::xlab("Library size (N reads)") +
        ggplot2::ylab("Effective library size \n(DESeq2 normalization factor)") +
		ggplot2::theme_classic() +
		ggplot2::theme(axis.text = ggplot2::element_text(size= 10), 
			axis.title = 	ggplot2::element_text(size = 8, face = "bold"),
			legend.title = ggplot2::element_text(size = 8, face = "bold")) +
		ggplot2::guides(size = FALSE)

	print(plot)
  
```

```{r DESeq2_conditional_block_ELS_text, eval="DESeq2" %in% names(all_counts_for_plotting), echo=FALSE, results='asis'}

cat("The effective library size is the factor used by DESeq2 normalization algorithm for each sample. The effective library size must be dependent of raw library size.")
```


```{r DESeq2_conditional_block_ma_head, eval='DESeq2' %in% names(all_counts_for_plotting), echo=FALSE, results='asis'}

cat("### **DESeq2 MA plot**:

This is the MA plot from DESeq2 package:")

```

```{r DESeq2_conditional_block_ma_plot, eval='DESeq2' %in% names(all_counts_for_plotting), echo=FALSE}
	DESeq2::plotMA(package_objects[['DESeq2']][['de_deseq2']], cex.lab=1.6, cex.axis=1.5)
```

```{r DESeq2_conditional_block_ma_text, eval='DESeq2' %in% names(all_counts_for_plotting), echo=FALSE, results='asis'}

	
cat("In DESeq2, the MA-plot (log ratio versus abundance) shows the log2 fold changes are attributable to a given variable over the mean of normalized counts. Points will be colored red if the adjusted Pvalue is less than 0.1. Points which fall out of the window are plotted as open triangles pointing either up or down.

A table containing the DESeq2 DEGs is provided: in Results\\_DESeq2/DEgenes\\_DESEq2.txt

A table containing the DESeq2 normalized counts is provided in Results\\_DESeq2/Normalized\\_counts\\_DESEq2.txt")

```

```{r DESeq2_only_prev_norm_head, eval="DESeq2" %in% names(all_counts_for_plotting), echo=FALSE, results='asis'}

cat("### **Differences between samples by PREVALENT DEGs normalized counts**:
		
Counts of prevalent DEGs were normalizated by DESeq2 algorithm. This count were scaled by log10 and plotted in a heatmap.")


```

```{r DESeq2_only_prev_norm_plot, eval="DESeq2" %in% names(all_counts_for_plotting), echo = FALSE, fig.height=12, fig.width=12}
	
	DE_prevalent_genes <- rownames(DE_all_genes[DE_all_genes[, "genes_tag"] %in% c("PREVALENT_DEG"),])
	if(length(DE_prevalent_genes) > 1){
		deseq_prevalent_normalized <- deseq_normalized_data[rownames(deseq_normalized_data) %in% DE_prevalent_genes,]  
		deseq_normalized_data_mtx <- log10(data.matrix(deseq_prevalent_normalized) + 1)
	    gplots::heatmap.2(deseq_normalized_data_mtx, col = col, labRow = FALSE, dendrogram = "column", trace = "none", srtCol=45,  adjCol = c(1,1))

	}else{
        cat("Lower than 2 prevalent differential expression were found")
	}
	
```


```{r edgeR_conditional_block_head, eval='edgeR' %in% names(all_counts_for_plotting), echo=FALSE, results='asis'}

cat("### **edgeR MA plot**

This is the MA plot from edgeR package:")

```

```{r edgeR_conditional_block_plot, eval='edgeR' %in% names(all_counts_for_plotting), echo=FALSE}

with(all_counts_for_plotting[['edgeR']], plot(logCPM, logFC, pch=20, main='edgeR: Fold change vs abundance', cex.lab=1.5, cex.axis=1.5))
with(subset(all_counts_for_plotting[['edgeR']], FDR < opt$p_val_cutoff), points(logCPM, logFC, pch=20, col='red'))
abline(h=c(-1,1), col='blue')
	
```

```{r edgeR_conditional_block_body, eval='edgeR' %in% names(all_counts_for_plotting), echo=FALSE, results='asis'}

cat("Differential gene expression data can be visualized as MA-plots (log ratio versus abundance) where each dot represents a gene. The differentially expressed genes are colored red and the non-differentially expressed ones are colored black.

A table containing the edgeR DEGs is provided in Results\\_edgeR/DEgenes\\_edgeR.txt

A table containing the edgeR normalized counts is provided in Results\\_edgeR/Normalized\\_counts\\_edgeR.txt")

```

```{r limma_conditional_block_head, eval='limma' %in% names(all_counts_for_plotting), echo=FALSE, results='asis'}

cat("### **limma Volcano plot**

Volcano plot of log2-fold change versus -log10 of adjusted p-values for all genes according to the analysis with limma:")

```

```{r limma_conditional_block_plot, eval='limma' %in% names(all_counts_for_plotting), echo=FALSE, warning=FALSE}

limma_sig <- row.names(all_counts_for_plotting[['limma']]) %in% row.names(DE_all_genes[DE_all_genes$limma_DEG == TRUE, ]) + 1
plot(x=all_counts_for_plotting[['limma']]$logFC, 
	y=-log10(all_counts_for_plotting[['limma']]$adj.P.Val),
	main= c('groupsB-groupsA'), xlab='logFC', ylab='-log10 adjusted p-value',
	col=c('blue', 'red')[limma_sig],
	pch=20,cex.lab=1.6, cex.axis=1.5)
abline(v= opt$lfc, col='cyan')	
abline(v= -(opt$lfc), col='cyan')	
abline(h=-log10(opt$p_val_cutoff), col='red', lty='dashed')

```

```{r limma_conditional_block_text, eval='limma' %in% names(all_counts_for_plotting), echo=FALSE, results='asis'}

cat("A table containing the limma DEGs is provided in Results\\_limma/DEgenes\\_limma.txt")


cat("A table containing the limma normalized counts is provided in Results\\_limma/Normalized\\_counts\\_limma.txt")
```

```{r NOISeq_conditional_block_head, eval='NOISeq' %in% names(all_counts_for_plotting), echo=FALSE, results='asis'}
cat("### **NOISeq Expressionplot**

This is the summary plot for (M,D) values (black) and the differentially expressed genes (red) from the NOISeq package (Image extracted from {'ExpressionPlot.pdf'} file):"

)
```

```{r NOISeq_conditional_block_plot, eval='NOISeq' %in% names(all_counts_for_plotting), echo=FALSE, warning=FALSE, message=FALSE, results='hide',}
    q_value <- 1-opt$p_val_cutoff # Differential expression threshold for creating plot
	NOISeq::DE.plot(package_objects[['NOISeq']], q = q_value, graphic = 'MD', cex.lab=1.4, cex.axis=1.4)

```

```{r NOISeq_conditional_block_text, eval='NOISeq' %in% names(all_counts_for_plotting), echo=FALSE, results='asis'}

cat("A table containing the NOISeq DEGs is provided in Results\\_NOISeq/DEgenes\\_NOISeq.txt.

A table containing the NOISeq normalized counts is provided in Results\\_NOISeq/Normalized\\_counts\\_NOISeq.txt"
	
)
```


```{r WGCNA_conditional_block_head, eval=grepl("W", opt$modules), echo=FALSE, results='asis'}
cat("## **WGCNA Results**

WGCNA was run to look for modules (clusters) of coexpressed genes. These modules were then compared with the sample factors to look for correlation. 
If no sample factors were specified, this comparison was performed with treatment/control labels.

The following graphic shows the power value chosen for building clusters. The power is chosen by looking at the characteristics of the network produced."

)
```

```{r WGCNA_conditional_block_plot_power_choice, eval=grepl("W", opt$modules), echo=FALSE, warning=FALSE, message=FALSE, fig.height=10, fig.width=15}
    replayPlot(results_WGCNA[['plot_objects']][['power_threshold_effects']])
```



```{r WGCNA_conditional_block_head_cluster_membership, eval=grepl("W", opt$modules), echo=FALSE, results='asis'}
cat("In total there were ", max(unique(results_WGCNA$gene_cluster_info$Cluster_ID)) , "clusters. 
	The following plot shows the number of genes per cluster:"
)
```

```{r WGCNA_conditional_block_plot_cluster_membership, eval=grepl("W", opt$modules), echo=FALSE, warning=FALSE, message=FALSE, fig.height=10, fig.width=15}
	genes_per_cluster <- table(results_WGCNA$gene_cluster_info$Cluster_ID)
	col_clusters_barplot <- results_WGCNA[['plot_objects']][['sorted_colours']]
	names(col_clusters_barplot) <- names(genes_per_cluster)
	ggplot2::ggplot(as.data.frame(genes_per_cluster), ggplot2::aes(Var1, Freq, fill=Var1)) + 
    ggplot2::geom_bar(stat="identity") + ggplot2::scale_fill_manual(name = "Cluster ID", values=col_clusters_barplot) + 
    ggplot2::labs(x = "Cluster ID (0 = unclustered)", y = "Frequency", main="Number of Genes per Cluster") + ggplot2::theme(legend.position = "none")
```

```{r WGCNA_conditional_dist_cluster_membership, eval=grepl("W", opt$modules), echo=FALSE, results='asis'}
cat("### Module Membership distribution\n"
)
```

```{r WGCNA_dist_cluster_membership, eval=grepl("W", opt$modules), echo=FALSE, warning=FALSE, message=FALSE, fig.height=10, fig.width=15}
		"%>%" <- magrittr::"%>%"

	mm_dist <- DE_all_genes[DE_all_genes$Cluster_ID != 0, c("Cluster_MM", "Cluster_ID")]
	mm_dist$Cluster_ID <- as.character(mm_dist$Cluster_ID)
	mm_summary <- mm_dist %>% 
					dplyr::group_by(Cluster_ID) %>% 
					dplyr::summarise(MM_sd = sd(Cluster_MM),
							MM_mean = mean(Cluster_MM),
							gene_count = length(Cluster_MM))
	mm_summary <- as.data.frame(mm_summary)				
	mm_tables_equivalence <- match(mm_dist$Cluster_ID, mm_summary$Cluster_ID)
	mm_dist <- cbind(mm_dist, mm_summary[mm_tables_equivalence, c("MM_sd", "MM_mean", "gene_count")])
	mm_dist <- mm_dist[!is.na(mm_dist$Cluster_ID),]		
	mm_dist$Cluster_ID <- factor(mm_dist$Cluster_ID, levels = as.character(sort(unique(as.numeric(mm_dist$Cluster_ID)))))
	# mm_dist<- mm_dist[order(as.numeric(mm_dist$Cluster_ID)),]	
	gg <- ggplot2::ggplot(mm_dist, ggplot2::aes(x = Cluster_ID, group = Cluster_ID, y = Cluster_MM, fill = gene_count)) + 
			ggplot2::geom_violin(trim = TRUE, scale = "width") + ggplot2::ylim(c(-0.00001, 1.00001)) +
  			ggplot2::geom_boxplot(width=0.05)+  
  			ggplot2::scale_fill_gradient(low = "#9FCBFC", high =  "#0067DC")
	gg

```



```{r WGCNA_conditional_block_plot_factors, eval=grepl("W", opt$modules), echo=FALSE, warning=FALSE, message=FALSE, fig.height=10, fig.width=15, results = 'asis'}
cat("### **Cluster assignment vs lower module membership (MM) **")

cat("\nThis plot shows, for each gene, the cluster ID ascertained by WGCNA, vs. the cluster whose eigen gene has the highest correlation (module membership/MM).")
	
	plot_counter <- plot_in_div(results_WGCNA[['plot_objects']][['cluster_vs_MM']], counter = plot_counter)
	cluster_count <- get_clusters_count(results_WGCNA)
	features_count <- get_features_count(results_WGCNA)
	
cat("### **Cluster vs. factors correlation**")

cat("\nThe following plots show the correlation between the different modules and specified factors. This is done using eigengenes, 
which can be broadly thought of as the average expression pattern for the genes in a given cluster. MEn refers to the eigengene for cluster n.\n")

cat("\nThis plot shows the correlation between clusters (eigen genes) and factors directly.")

	pp <- gg_heatmap(data =  results_WGCNA[['plot_objects']][['trait_vs_module']], 
						x_axis = "Trait",
						y_axis= "Module",
						fill = "Correlation",
						text_plot= "Text_correlation")
	plot_counter <- plot_in_div(pp, fig_height= (5 + (0.8 * cluster_count)), fig_width=(3 + (features_count * 0.8)), cex = 0.5, counter = plot_counter)

cat("### **WGCNA Eigen values clustering**")

cat("\nWGCNA dendogram showing distances between these eigengenes along with the factors. Distances has been calculated using **signed** correlation so more near elements, more positive correlation between elements.") 
	
	plot_counter <- plot_in_div(WGCNA::plotEigengeneNetworks(results_WGCNA[['plot_objects']][['trait_and_module']], "Eigengene dendrogram", marDendro = c(0,4,2,0), plotHeatmaps = FALSE),
		fig_width= 4+ calc_width_clusters((features_count + cluster_count), multiplier = 0.15),  min_size = 5, counter = plot_counter)

cat("### **Eigen values clustering (Absolute correlation)**")

cat("\nWGCNA like dendogram showing distances between these eigengenes along with the factors. Distances has been calculated using **absolute** correlation so more near elements, more absolute correlation between elements.")

	corME <- cor(results_WGCNA[['plot_objects']][['trait_and_module']], use="p")
	disME_abs <- as.dist(1-abs(corME))
	clust_abs = fastcluster::hclust(disME_abs, method = "average")
	
	plot_counter <- plot_in_div(plot(clust_abs, sub="", xlab="", ylab="", ylim=c(0,1)),
		fig_width= 4+ calc_width_clusters((features_count + cluster_count), multiplier = 0.15),  min_size = 5, counter = plot_counter)


corME2 <- corME
rownames(corME2) <- gsub("^ME", "", rownames(corME2))
colnames(corME2) <- gsub("^ME", "", colnames(corME2))

mod_traits_net <- corM2igraph(corME2, cor_abs_thr = 0.8)

mod_net <- igraph::graph_from_data_frame(d=mod_traits_net$edges, vertices = mod_traits_net$nodes, directed = FALSE)
# layout <- igraph::layout_with_fr(mod_net, weights = mod_traits_net$edges$corr * 3 )

cat("\n### **Correlation network between modules and factors**\n\n")

if (nrow(mod_traits_net$nodes < 2)){

	cat("This plot shows modules (black) and factors (green) as nodes. Correlations coefficients over 0.8 (red) and under -0.8 (blue) are represented as edges\n")
	"%>%" <- magrittr::"%>%"
	options(ggrepel.max.overlaps = Inf)
	tg <- tidygraph::as_tbl_graph(mod_net) %>% tidygraph::activate(nodes) %>% dplyr::mutate(label=name)
	igraph::E(tg)$weight <- igraph::E(tg)$corr
	test_graph <- tg %>% 
	ggraph::ggraph(layout = "stress")+  
	ggraph::geom_edge_fan(ggplot2::aes(colour = ifelse(mod_traits_net$edges$corr_type == "corr", "#FF7272","#9595FF"))) +
	ggraph::geom_node_text(ggplot2::aes(label = name), 
	       repel = TRUE, point.padding  = grid::unit(0.2, "lines"), 
	       colour=ifelse(grepl("_",mod_traits_net$nodes$nodes), "#43CF4C", "black"))+
	ggraph::theme_graph(background = 'white')


	plot_counter <- plot_in_div(test_graph)
	options(ggrepel.max.overlaps = 10)

} else {
	cat("Only 2 modules or factors correlates with higher ansolute correlation coefficient than 0.85.")
}


cat("\n### **Correlation between all clusters and factors**\n")

	correlation <- cor(results_WGCNA[['plot_objects']][['trait_and_module']], use = "p")
	correlation_df <- as.data.frame(correlation)
	pp <- gg_heatmap(data_table = correlation_df, 
    						fill = "Correlation",
    						labs = FALSE,
    						input = "matrix",
    						x_angle = 25)
	cor_elements <- features_count + cluster_count
	plot_counter <- plot_in_div(pp,	5 + (cor_elements *0.2), 5 + (cor_elements *0.2), counter = plot_counter)

```

```{r diffcoexp_conditional_block_head, eval=grepl("X", opt$modules), echo=FALSE, results='asis'}
cat("### **diffcoexp Results**

The package diffcoexp looks for genes that are coexpressed with different sets genes for case and control samples respectively, i.e. differentially co-expressed

The following plot shows the ratio of differentially co-expressed links (DCLs) to the total number of co-expressed links for each gene for which at least one DCL is available:")
```

```{r diffcoexp_conditional_block_plot, eval=grepl("X", opt$modules), echo=FALSE, warning=FALSE, message=FALSE}
DCGs <- results_diffcoexp$DCGs
DCG_plot_df <-data.frame(DCG_ratio = DCGs$DCLs/DCGs$CLs, 
                mean_logFCs = DE_all_genes[DCGs$Gene,"mean_logFCs"],
                DCG_qval = DCGs$q,
                combined_FDR = DE_all_genes[DCGs$Gene,"combined_FDR"])

# Use these to vary signficance according to DCG and combiend FDR for DE
DCG_plot_df$DCG_sig <- ifelse(DCG_plot_df$DCG_qval < 0.05, "sig", "not_sig")
DCG_plot_df$DE_sig <- ifelse(DCG_plot_df$combined_FDR < 0.05, "sig", "not_sig")
ggplot2::ggplot(DCG_plot_df, ggplot2::aes(x=DCG_ratio, y=mean_logFCs, colour=DCG_sig, shape=DE_sig)) + ggplot2::geom_point()

```


```{r PCIT_conditional_block_head, eval=!is.null(DE_all_genes[['rif1']]), echo=FALSE, results='asis'}
cat("### **PCIT Results**

Plots of some of the various metrics computed in the PCIT implementation. Better explanation and more plots to be added at a later date:"

)
```

```{r PCIT_conditional_block_plot, eval=!is.null(DE_all_genes[['rif1']]), echo=FALSE, warning=FALSE, message=FALSE}
    ggplot2::ggplot(DE_all_genes, ggplot2::aes(x=ctrl_cn)) +
              ggplot2::geom_density()
    ggplot2::ggplot(DE_all_genes, ggplot2::aes(x=treat_cn)) +
              ggplot2::geom_density()
    ggplot2::ggplot(DE_all_genes, ggplot2::aes(x=diff_cn)) +
            ggplot2::geom_density()
    ggplot2::ggplot(DE_all_genes, ggplot2::aes(x=diffK)) +
            ggplot2::geom_density()

    av_de <- ggplot2::ggplot(DE_all_genes, ggplot2::aes(x=average, y=de)) +
            ggplot2::geom_point()
    av_de <- ggExtra::ggMarginal(
      av_de,
      type = 'density',
      margins = 'both',
      size = 5,
      colour = '#000000',
      fill = '#A6A39E'
    )

    de_diffK <- ggplot2::ggplot(DE_all_genes, ggplot2::aes(x=de, y=diffK)) +
            ggplot2::geom_point()
    de_diffK <- ggExtra::ggMarginal(
      de_diffK,
      type = 'density',
      margins = 'both',
      size = 5,
      colour = '#000000',
      fill = '#A6A39E'
    )

    de_pif <- ggplot2::ggplot(DE_all_genes, ggplot2::aes(x=de, y=pif)) +
            ggplot2::geom_point()
    de_pif <- ggExtra::ggMarginal(
      de_pif,
      type = 'density',
      margins = 'both',
      size = 5,
      colour = '#000000',
      fill = '#A6A39E'
    )

    de_rif1 <- ggplot2::ggplot(DE_all_genes, ggplot2::aes(x=de, y=rif1)) +
            ggplot2::geom_point()
    de_rif1 <- ggExtra::ggMarginal(
      de_rif1,
      type = 'density',
      margins = 'both',
      size = 5,
      colour = '#000000',
      fill = '#A6A39E'
    )

    de_rif2 <- ggplot2::ggplot(DE_all_genes, ggplot2::aes(x=de, y=rif2)) +
            ggplot2::geom_point()
    de_rif2 <- ggExtra::ggMarginal(
      de_rif2,
      type = 'density',
      margins = 'both',
      size = 5,
      colour = '#000000',
      fill = '#A6A39E'
    )

    rif1_rif2 <- ggplot2::ggplot(DE_all_genes, ggplot2::aes(x=rif1, y=rif2)) +
            ggplot2::geom_point()
    rif1_rif2 <- ggExtra::ggMarginal(
      rif1_rif2,
      type = 'density',
      margins = 'both',
      size = 5,
      colour = '#000000',
      fill = '#A6A39E'
    )
    gridExtra::grid.arrange(av_de, de_diffK, de_pif, de_rif1, de_rif2, rif1_rif2)

```

## **Detailed package results comparation**
This is an advanced section in order to compare the output of the packages used to perform data analysis. The data shown here does not necessarilly have any biological implication. 

### **P-value Distributions**
Distributions of p-values, unadjusted and adjusted for multiple testing (FDR)

```{r pval_distr, echo = FALSE, warning = FALSE}

df <- DE_all_genes[DE_all_genes$genes_tag != "FILTERED_OUT", ]
df$gene_names <- row.names(df)
df_melted_pval <- reshape2::melt(df, id.vars = "gene_names", measure.vars = final_FDR_names)
df_melted_FDR <- reshape2::melt(df, id.vars = "gene_names", measure.vars = final_pvalue_names)
# Now plot
ggplot2::ggplot(df_melted_pval, ggplot2::aes(x=value)) + 
    ggplot2::geom_density() + ggplot2::facet_grid(. ~variable)
ggplot2::ggplot(df_melted_FDR, ggplot2::aes(x=value)) + 
    ggplot2::geom_density() + ggplot2::facet_grid(. ~variable)
```

### **FDR Correlations**
Correlations of adjusted p-values, adjusted for multiple testing (FDR) and for log Fold Change.

```{r pval_corr, echo = FALSE, warning = FALSE, eval = length(all_counts_for_plotting) > 1}

PerformanceAnalytics::chart.Correlation(df[, c("combined_FDR", final_FDR_names)], histogram=TRUE, pch=19, log="xy", na.action = na.omit)
PerformanceAnalytics::chart.Correlation(df[, final_pvalue_names], histogram=TRUE, pch=19, log="xy", na.action = na.omit)
PerformanceAnalytics::chart.Correlation(df[, c("mean_logFCs", final_logFC_names)], histogram=TRUE, pch=19, log="xy", na.action = na.omit)

```

### **Values of options used to run DEGenesHunter**
First column contains the option names; second column contains the given values for each option in this run.
```{r opt_vals, echo = FALSE, warning = FALSE}
knitr::kable(cbind(opt))
```
